Self test compilation fails with GCC 12.1.0

### Summary

With GCC 12.1.0, compilation of selftest.c fails (this was found running `all.sh test_no_platform`

```
test/selftest.c: In function ‘calloc_self_test’:
test/selftest.c:129:17: error: pointer ‘buffer1’ may be used after ‘free’ [-Werror=use-after-free]
  129 |     old_buffer1 = (uintptr_t) buffer1;
      |     ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
test/selftest.c:63:28: note: call to ‘free’ here
   63 | #define mbedtls_free       free
      |                            ^
test/selftest.c:130:5: note: in expansion of macro ‘mbedtls_free’
  130 |     mbedtls_free( buffer1 );
      |     ^~~~~~~~~~~~
```
This is (partially) technically correct, old_buffer1 retains a pointer to the then freed block of memory, although nothing is then done with it other than checking its value against the new returned pointer.

This is a new warning as of GCC 12, and turned on as part of -Wall, so I created this issue for discussion. We are not the only project having issues like this, and I can find virtually no documentation on the new warning.

I am not really sure how best to get around this, to be quite honest. I am not keen on suppressing this warning as I haven't seen any other failures as yet because of this, and its a potentially extremely useful feature.

--- a/ChangeLog.d/selftest-gcc12.txt
+++ b/ChangeLog.d/selftest-gcc12.txt
@@ -1,0 +1,2 @@
+Bugfix
+   * Silence a warning from GCC 12 in the selftest program. Fixes #5974.

--- a/programs/test/selftest.c
+++ b/programs/test/selftest.c
@@ -84,7 +84,6 @@ static int calloc_self_test( int verbose )
     void *empty2 = mbedtls_calloc( 0, 1 );
     void *buffer1 = mbedtls_calloc( 1, 1 );
     void *buffer2 = mbedtls_calloc( 1, 1 );
-    uintptr_t old_buffer1;
 
     if( empty1 == NULL && empty2 == NULL )
     {
@@ -126,7 +125,6 @@ static int calloc_self_test( int verbose )
             mbedtls_printf( "  CALLOC(1): passed\n" );
     }
 
-    old_buffer1 = (uintptr_t) buffer1;
     mbedtls_free( buffer1 );
     buffer1 = mbedtls_calloc( 1, 1 );
     if( buffer1 == NULL )
@@ -138,9 +136,7 @@ static int calloc_self_test( int verbose )
     else
     {
         if( verbose )
-            mbedtls_printf( "  CALLOC(1 again): passed (%s address)\n",
-                            (uintptr_t) old_buffer1 == (uintptr_t) buffer1 ?
-                            "same" : "different" );
+            mbedtls_printf( "  CALLOC(1 again): passed\n" );
     }
 
     if( verbose )

